### PR TITLE
 Improve testutils.MarshalAny

### PR DESCRIPTION
**Pull request by ulascansenturk**
- Used anypb.New() instead of ptypes.MarshalAny

fixes: #6611 

RELEASE NOTES: N/A

### CODE REVIEW

File: internal/status/status.go [hunk 1]
```diff
@@ -109,7 +111,7 @@ func (s *Status) WithDetails(details ...proto.Message) (*Status, error) {
 	// s.Code() != OK implies that s.Proto() != nil.
 	p := s.Proto()
 	for _, detail := range details {
-		any, err := ptypes.MarshalAny(detail)
+		any, err := anypb.New(protoadapt.MessageV2Of(detail))
```
Comment (easwars) [inline internal/status/status.go:114]: Please keep this PR to only `testutils` changes and its callsites.

File: internal/testutils/marshal_any.go [hunk 2]
```diff
@@ -19,18 +19,30 @@ package testutils
...
 	if err != nil {
-		panic(fmt.Sprintf("ptypes.MarshalAny(%+v) failed: %v", m, err))
+		return nil, fmt.Errorf("anypb.New(%+v) failed: %w", m, err)
+	}
+	return a, nil
+}
+
+// TestMarshalAny is a wrapper function for testing. It marks itself as a test helper.
+// It uses MarshalAny and calls t.Fatalf if marshaling fails.
+func TestMarshalAny(t *testing.T, m proto.Message) *anypb.Any {
```
Comment (easwars) [inline internal/testutils/marshal_any.go:41]: Why do we need two functions here?

File: internal/testutils/marshal_any.go [hunk 3]
```diff
@@ -19,18 +19,30 @@ package testutils
...
 	"github.com/golang/protobuf/proto"
-	"github.com/golang/protobuf/ptypes"
 	"google.golang.org/protobuf/types/known/anypb"
 )
 
 // MarshalAny is a convenience function to marshal protobuf messages into any
 // protos. It will panic if the marshaling fails.
-func MarshalAny(m proto.Message) *anypb.Any {
-	a, err := ptypes.MarshalAny(m)
+func MarshalAny(m proto.Message) (*anypb.Any, error) {
```
Comment (easwars) [inline internal/testutils/marshal_any.go:31]: Why does this have to return an error? This should be changed to accept a `testing.T` and directly call `t.Fatal`.

**Review by easwars — CHANGES_REQUESTED** [review]

File: internal/testutils/marshal_any.go (hunk 2 above)
Comment (ulascansenturk) [reply internal/testutils/marshal_any.go:41]: Because testutils.MarshalAny was not only used in tests, for example it was also used in the clientresources package, I thought it would be a good idea to create a new function, but I will revert it

File: internal/testutils/marshal_any.go (hunk 3 above)
Comment (ulascansenturk) [reply internal/testutils/marshal_any.go]: Done

File: internal/testutils/marshal_any.go (hunk 2 above)
Comment (easwars) [reply internal/testutils/marshal_any.go]: The `clientresources` package is inside of `internal/testutils` and hence is also a test package.

File: internal/testutils/marshal_any.go [hunk 4]
```diff
@@ -18,19 +18,20 @@
...
 	"google.golang.org/protobuf/types/known/anypb"
+	"testing"
 )
 
 // MarshalAny is a convenience function to marshal protobuf messages into any
 // protos. It will panic if the marshaling fails.
-func MarshalAny(m proto.Message) *anypb.Any {
-	a, err := ptypes.MarshalAny(m)
+func MarshalAny(t *testing.T, m proto.Message) *anypb.Any {
+	t.Helper() // Mark this function as a test helper.
```
Comment (easwars) [inline internal/testutils/marshal_any.go:30]: Please remove that trailing comment.

File: internal/testutils/marshal_any.go [hunk 5]
```diff
@@ -18,19 +18,20 @@
...
+
+	a, err := anypb.New(protoadapt.MessageV2Of(m))
 	if err != nil {
-		panic(fmt.Sprintf("ptypes.MarshalAny(%+v) failed: %v", m, err))
+		t.Fatalf("ptypes.MarshalAny(%+v) failed: %v", m, err) // Use t.Fatalf instead of panic.
```
Comment (easwars) [inline internal/testutils/marshal_any.go:34]: Same here. Please remove the trailing comment.

File: internal/testutils/marshal_any.go (hunk 2 above)
Comment (ulascansenturk) [reply internal/testutils/marshal_any.go]: reverted, thanks for the heads up

File: internal/testutils/marshal_any.go (hunk 5 above)
Comment (easwars) [inline internal/testutils/marshal_any.go:34]: Also, please change the error string.

`t.Fatalf("Failed to marshal proto %+v into an Any: %v", m, err)`

File: internal/testutils/xds/e2e/clientresources.go [hunk 6]
```diff
@@ -106,7 +107,8 @@ var RouterHTTPFilter = HTTPFilter("router", &v3routerpb.Router{})
 // DefaultClientListener returns a basic xds Listener resource to be used on
 // the client side.
 func DefaultClientListener(target, routeName string) *v3listenerpb.Listener {
-	hcm := testutils.MarshalAny(&v3httppb.HttpConnectionManager{
+
```
Comment (easwars) [inline internal/testutils/xds/e2e/clientresources.go:110]: Please remove this newline.

File: interop/xds/custom_lb_test.go [hunk 7]
```diff
@@ -66,7 +66,7 @@ func (s) TestCustomLB(t *testing.T) {
 			}
 			rpcBMD := md.Get("rpc-behavior")
 			if len(rpcBMD) != 1 {
-				errCh.Send(fmt.Errorf("received %d values for metadata key rpc-behavior, want 1", len(rpcBMD)))
```
Comment (easwars) [inline interop/xds/custom_lb_test.go:69]: Please undo this change.

File: status/status_test.go [hunk 8]
```diff
@@ -22,10 +22,10 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"github.com/golang/protobuf/ptypes"
```
Comment (easwars) [inline status/status_test.go:25]: Please revert this change.

File: xds/internal/clusterspecifier/rls/rls_test.go [hunk 9]
```diff
@@ -107,56 +108,64 @@ func (s) TestParseClusterSpecifierConfig(t *testing.T) {
...
 // This will error because the required match field is set in grpc key builder.
-var rlsClusterSpecifierConfigError = testutils.MarshalAny(&rlspb.RouteLookupClusterSpecifier{
-	RouteLookupConfig: &rlspb.RouteLookupConfig{
-		GrpcKeybuilders: []*rlspb.GrpcKeyBuilder{
-			{
-				Names: []*rlspb.GrpcKeyBuilder_Name{
-					{
-						Service: "service",
-						Method:  "method",
+func rlsClusterSpecifierConfigError(t *testing.T) *anypb.Any {
```
Comment (easwars) [inline xds/internal/clusterspecifier/rls/rls_test.go:111]: Given that vars `rlsClusterSpecifierConfigError` and `rlsClusterSpecifierConfigWithoutTransformations` are only ever referenced from the test table above, please inline them instead of making a function for them.

File: xds/internal/resolver/xds_resolver_test.go [hunk 10]
```diff
@@ -83,7 +83,7 @@ var (
 	target = resolver.Target{URL: *testutils.MustParseURL("xds:///" + targetStr)}
 
 	routerHTTPFilter = httpfilter.Get(router.TypeURL)
-	routerConfig, _  = routerHTTPFilter.ParseFilterConfig(testutils.MarshalAny(&v3routerpb.Router{}))
+	routerConfig, _  = routerHTTPFilter.ParseFilterConfig(testutils.MarshalAny(&testing.T{}, &v3routerpb.Router{}))
```
Comment (easwars) [inline xds/internal/resolver/xds_resolver_test.go:86]: Please add a function instead of these global vars. That way you would be able to pass a valid `testing.T` to `testutils.MarshalAny`.

```golang
func makeRouterFilter(t *testing.T) xdsresource.HTTPFilter {
	f := httpfilter.Get(router.TypeURL)
	cfg, _  = f.ParseFilterConfig(testutils.MarshalAny(t, &v3routerpb.Router{}))
	return xdsresource.HTTPFilter{Name: "rtr", Filter: routerHTTPFilter, Config: cfg}
}
```

File: xds/internal/server/listener_wrapper_test.go [hunk 11]
```diff
@@ -51,111 +51,118 @@ const (
...
-	FilterChains: []*v3listenerpb.FilterChain{
-		{
-			FilterChainMatch: &v3listenerpb.FilterChainMatch{
-				PrefixRanges: []*v3corepb.CidrRange{
-					{
-						AddressPrefix: "192.168.0.0",
-						PrefixLen: &wrapperspb.UInt32Value{
-							Value: uint32(16),
+func listenerWithRouteConfiguration(t *testing.T) *v3listenerpb.Listener {
+
```
Comment (easwars) [inline xds/internal/server/listener_wrapper_test.go:55]: Remove this and the next unnecessary newlines.

File: xds/internal/xdsclient/xdsresource/filter_chain_test.go [hunk 12]
```diff
@@ -91,7 +95,7 @@ var (
 	}
 	emptyRouterFilter = e2e.RouterHTTPFilter
 	routerBuilder     = httpfilter.Get(router.TypeURL)
-	routerConfig, _   = routerBuilder.ParseFilterConfig(testutils.MarshalAny(&v3routerpb.Router{}))
+	routerConfig, _   = routerBuilder.ParseFilterConfig(testutils.MarshalAny(&testing.T{}, &v3routerpb.Router{}))
```
Comment (easwars) [inline xds/internal/xdsclient/xdsresource/filter_chain_test.go:98]: Similar to the comment about the routerFilter in the other test, consider making a function instead.

File: xds/internal/xdsclient/xdsresource/unmarshal_lds_test.go [hunk 13]
```diff
@@ -1852,17 +1852,18 @@ var customFilterNewTypedStructConfig = &v3xdsxdstypepb.TypedStruct{
 var wrappedCustomFilterNewTypedStructConfig *anypb.Any
 
 func init() {
-	wrappedCustomFilterOldTypedStructConfig = testutils.MarshalAny(customFilterOldTypedStructConfig)
-	wrappedCustomFilterNewTypedStructConfig = testutils.MarshalAny(customFilterNewTypedStructConfig)
+	t := &testing.T{}
```
Comment (easwars) [inline xds/internal/xdsclient/xdsresource/unmarshal_lds_test.go:1855]: The whole idea of passing a `testing.T` to `testutils.MarshalAny` is to be able to fail that particular test when something errors out. This concept of passing an empty `testing.T` is not going to work.

File: internal/testutils/xds/e2e/clientresources.go [hunk 14]
```diff
@@ -159,14 +162,54 @@ func DefaultServerListener(host string, port uint32, secLevel SecurityLevel) *v3
...
 		ts = &v3corepb.TransportSocket{
 			Name: "envoy.transport_sockets.tls",
 			ConfigType: &v3corepb.TransportSocket_TypedConfig{
-				TypedConfig: testutils.MarshalAny(tlsContext),
+				TypedConfig: testutils.MarshalAny(&testing.T{}, tlsContext),
 			},
 		}
 	}
+
+	filterTypedConfig := testutils.MarshalAny(&testing.T{}, &v3httppb.HttpConnectionManager{
```
Comment (easwars) [inline internal/testutils/xds/e2e/clientresources.go:175]: Is there any difference between `filterTypedConfig` and `filteredTypedConfig`? Also passing an empty `testing.T` is not what we want.

Comment (easwars) [reply internal/testutils/xds/e2e/clientresources.go:175]: It might be too painful to convert functions in this package to accept a `testing.T` that can be forwarded along `testutils.MarshalAny`. Instead you could have an unexported function in this function which resembles `panic`s instead of calling `t.Fatal` (like the old implementation).

**Review by easwars — CHANGES_REQUESTED** [review]

**Review by bitcoinfinancier — APPROVED** [review]

**Comment by easwars** [comment]
Could you please look at the `vet` failure? Thanks.

File: internal/testutils/marshal_any.go [hunk 15]
```diff
@@ -18,19 +18,20 @@
...
-
 	"github.com/golang/protobuf/proto"
-	"github.com/golang/protobuf/ptypes"
+	"google.golang.org/protobuf/protoadapt"
 	"google.golang.org/protobuf/types/known/anypb"
+	"testing"
 )
 
 // MarshalAny is a convenience function to marshal protobuf messages into any
 // protos. It will panic if the marshaling fails.
```
Comment (easwars) [inline internal/testutils/marshal_any.go:28]: This comment should no longer say that it will panic.

File: xds/internal/clusterspecifier/rls/rls_test.go [hunk 16]
```diff
@@ -107,56 +152,9 @@ func (s) TestParseClusterSpecifierConfig(t *testing.T) {
 }
 
 // This will error because the required match field is set in grpc key builder.
```
Comment (easwars) [inline xds/internal/clusterspecifier/rls/rls_test.go:154]: These comments need to go. They are orphaned now.

File: xds/internal/clusterspecifier/rls/rls_test.go [hunk 17]
```diff
@@ -57,13 +57,58 @@ func (s) TestParseClusterSpecifierConfig(t *testing.T) {
 		wantErr    bool
 	}{
 		{
-			name:    "invalid-rls-cluster-specifier",
-			rlcs:    rlsClusterSpecifierConfigError,
+			name: "invalid-rls-cluster-specifier",
```
Comment (easwars) [inline xds/internal/clusterspecifier/rls/rls_test.go:60]: Please add the comment here:

`// This will error because the required_match field is set in grpc key builder`

Comment (ulascansenturk) [reply xds/internal/clusterspecifier/rls/rls_test.go:60]: All done

**Comment by ulascansenturk** [comment]
> Could you please look at the `vet` failure? Thanks.

fixed

**Review by easwars — APPROVED** [review]

**Review by dfawley — APPROVED** [review]

**Comment by easwars** [comment]
Thanks @ulascansenturk, for your contribution!!
